### ISSUE TITLE
APIs to get, set and remove localStorage just like cookies

### ISSUE THREAD

**Issue by cuixiping** [electron/electron#9110]
I expect some apis to get, set and remove localStorage data, just like those apis of cookies.

Here are apis of cookies:
```
ses.cookies.get(filter, callback)
ses.cookies.set(details, callback)
ses.cookies.remove(url, name, callback)
```

Expected apis of localStorage
```
ses.localStorage.get(filter, callback)
ses.localStorage.set(details, callback)
ses.localStorage.remove(url, name, callback)
```

**Comment by deepak1556** [comment]
Can you explain the use case for this api ? Thanks!

**Comment by assetcorp** [comment]
What exactly do you need to accomplish. Are you going to use the apis in the main or renderer view or both?

**Comment by cuixiping** [comment]
@assetcorp use in the main process

**Comment by deepak1556** [comment]
@cuixiping can you explain the reason you require this api in the main process, what problem does it solve ? Unlike cookies api which are needed for network related apis in the main process I don't see the requirement for localstorage.

**Comment by cuixiping** [comment]
I am writing an app depending on `localStorage` . I hope there is a simple way to let the main process read and change localStorage data of specified origins, so I can easily backup and restore any items of localStorage without opening a window and loading a page.

There is an api for clearing localStorage but no for reading and changing.
[ses.clearStorageData([options, callback])](https://electron.atom.io/docs/api/session/#sesclearstoragedataoptions-callback)

**Comment by pabloleban** [comment]
You can get something from the localstorage in your main with
`mainWindow.webContents.executeJavaScript('localStorage.getItem("my-item")').then(function(value){console.log("my-item data: "+value)});`

**Comment by bpasero** [comment]
This is not really my scenario, I would like to access all of `localStorage` before even opening a window.

**Comment by fttx** [comment]
+1

**Comment by zhouzr** [comment]
+1

**Comment by csaund** [comment]
It seems so odd that the main process wouldn't have access to `localStorage`. I have a similar use case (would like to alter the main BrowserWindow before launch based on user history), trying to dig through implementation of `localStorage` to see if I can do something similar through `app.getPath('userData')`, but that feels like barking up the wrong tree. 

It looks like this is possible with some external libraries like [electron-json-storage](https://www.npmjs.com/package/electron-json-storage). Otherwise I suppose we should just change all instances of `localStorage` within the bowser process to use this package, but duplicate it for webapps?

**Comment by stanleyxu2005** [comment]
I have a concrete scenario: I want to migrate a web application into electron without code change. The existing web application uses localStorage to store some data. but window.localStorage, window.sessionStorage are not accessible in a Electron browser window. We need a solution for this.

**Comment by devqazi** [comment]
@csaund  Why dont you do it via nodes fs. Just store whatever attributes you want into a JSON file in `app.getPath('userData') + '/config.json'
`

**Comment by jerry1100** [comment]
FWIW, I recently ran into this problem as well and published [electron-browser-storage](https://www.npmjs.com/package/electron-browser-storage) to enable me to access `localStorage` (and `sessionStorage`) from the main process.

Example:
```javascript
// In main process
const { localStorage } = require('electron-browser-storage');
 
await localStorage.setItem('favorite_number', 12);
await localStorage.getItem('favorite_number'); // '12'
```

**Comment by Praveer1981** [comment]
@jerry1100  It does not work for me.
here is the sample code.
`app.on("ready", () => {
    createWindow();
    await localStorage.setItem('favorite_number', 12);
})
`
Error:

> [Window Title]
Error

[Main Instruction]
A JavaScript error occurred in the main process

[Content]
Uncaught Exception:
G:\Technical\Praveer_practice\ElectronProject\IPC\main.js:69
    await localStorage.setItem('favorite_number', 12);
    ^^^^^

SyntaxError: await is only valid in async function
    at Module._compile (internal/modules/cjs/loader.js:752:23)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:827:10)
    at Module.load (internal/modules/cjs/loader.js:677:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:609:12)
    at Function.Module._load (internal/modules/cjs/loader.js:601:3)
    at loadApplicationPackage (G:\Technical\Praveer_practice\ElectronProject\IPC\node_modules\electron\dist\resources\default_app.asar\main.js:107:12)
    at Object.<anonymous> (G:\Technical\Praveer_practice\ElectronProject\IPC\node_modules\electron\dist\resources\default_app.asar\main.js:155:5)
    at Module._compile (internal/modules/cjs/loader.js:815:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:827:10)
    at Module.load (internal/modules/cjs/loader.js:677:32)

[OK]

**Comment by jerry1100** [comment]
@Praveer1981 you need to declare the callback as an `async` function, as so:

```javascript
```

**Comment by Praveer1981** [comment]
ohh, I see. 
Is it possible to use `window.localstorage.getitem("favorite_number")`  in the render process.
I want to set in Main process and want to get in render process.

**Comment by jerry1100** [comment]
@Praveer1981 yes, that's exactly how you'd do it.

Make sure however that the renderer process is loaded using `file://` or else it won't be able to access the same storage, since the storage is domain-specific.

**Comment by Praveer1981** [comment]
**main.js**
`
const localStorage = require('electron-browser-storage');
app.on("ready", async () => {
    createWindow();
    await localStorage.setItem('favorite_number', 12);
})`
**render.js**
`const errorButton = document.getElementById('btnErrorButton');
errorButton.addEventListener('click', () =>{
    const data = window.localStorage.getItem('favorite_number');
    alert(`${data}`);
})`
Is it correct ?

**Comment by jerry1100** [comment]
Yeah that should work

**Comment by Praveer1981** [comment]
No, It didnt work for me. Do you have any test app ?

**Comment by jerry1100** [comment]
Are you using `BrowserWindow.loadFile()`, or `BrowserWindow.loadURL('file://...')`?

Again, this only works if you're loading a **file using the file:// protocol**, so something like `BrowserWindow.loadUrl('http://localhost:3000')` **won't work.**

**Comment by Praveer1981** [comment]
I used
mainWindow.loadURL('file:///index.html');

I will create a sample app and share it with you. I will share the github
link.

**Comment by cuixiping** [comment]
> FWIW, I recently ran into this problem as well and published [electron-browser-storage](https://www.npmjs.com/package/electron-browser-storage) to enable me to access `localStorage` (and `sessionStorage`) from the main process.
> 
> Example:
> 
> ```js
> // In main process
> const { localStorage } = require('electron-browser-storage');
>  
> await localStorage.setItem('favorite_number', 12);
> await localStorage.getItem('favorite_number'); // '12'
> ```

How to access a domain's storage in main process?

**Comment by Praveer1981** [comment]
@cuixiping   What do you mean by domain’s storage ?

**Comment by cuixiping** [comment]
> @cuixiping What do you mean by domain’s storage ?

Each storage in renderer process is origin related, for example "https://google.com" has it's own storage, how to access this storage in main process?

**Comment by jerry1100** [comment]
@cuixiping this isn't supported and I don't have plans to support this use case. I feel that it's more trouble than it's worth.

There are some issues with loading a remote resource just to get access to the origin's storage. To name a few: 1) if the user doesn't have internet, 2) if the resource redirects you to a _different_ origin, 3) if the resource is down. These will all break the API or create unintentional behavior.

Most apps should serve their content from a local file. If you need to display remote content, consider a `webview` or `iframe`.

**Comment by cuixiping** [comment]
> remote resource

It's not a remote resource. It looks like cookies very much.

**Comment by Prinzhorn** [comment]
I'm using `localStorage` + `BroadcastChannel` to synchronize my svelte store (could be redux or whatever you use) across all windows. This works perfectly and allows me to quickly prototype outside of Electron with plain Chrome. It does work perfectly fine inside Electron as well, but there I need access to certain stores inside the main process as well (e.g. one boolean value corresponds to a `MenuItem` of type checkbox in the top menu). `localStorage` is my single source of truth, but since there is no API to access it from within the main process I'd need to have a second store and keep them in sync (painful, source of bugs).

As others have pointed out there is already a method on the Session object to interact with storage (`ses.clearStorageData([options])`) and it would kind of make sense to expose more fine grained APIs as well. There are userland modules such as [electron-store](https://github.com/sindresorhus/electron-store) but the point is that Chromium already comes with a capable key/value store that I can easily use inside the render process, even with `nodeIntegration` disabled (which it should be).

**Comment by Prinzhorn** [comment]
Maybe we can get more practical here: can someone with knowledge of Chromium internals explain how access to LevelDB is managed? LevelDB can only be used by a single process. In a simple experiment I can see that launching only the Electron main process without opening a BrowserWindow does not touch the files inside `~/.config/electron-quick-start/Local Storage/leveldb`. Opening a window (which does not interact with `localStorage` at all) causes access to the filesystem. Now the question is how does Chromium synchronize LevelDB access across multiple renderer processes? Is the main process involved at all? I'm asking this because I was thinking about using a LevelDB client to access `localStorage` directly from the main process (essentially https://github.com/regular/chrome-localstorage/blob/63d7e982d9faf7c6a3219101af530f2f5cbc6fa7/index.js). But like I said LevelDB can only be used by a single process by design so either the main process won't get access due to locks (or it locks the renderer out) or might end up with inconsistent data.

TLDR: is the main process currently involved in `localStorage` (LevelDB) business at all? And can we expose an API to users?

**Comment by mahnunchik** [comment]
Any news?

**Comment by fabiospampinato** [comment]
Echoing on what others have said: this would also be super useful for passing some data to new BrowserWindows in a way that's immediately available to them as soon as they start executing any code.

**Comment by ksm2** [comment]
@fabiospampinato Could you not achieve the same thing using a preload script? Then pass the value you need available by the renderer process by using the `webPreferences.additionalArguments`:

```js
// main.js
const browserWindow = new BrowserWindow({
  webPreferences: {
    preload: 'preload.js',
    additionalArguments: ['key=value'],
  },
});

// preload.js
const [key, value] = process.argv[0].split('=', 2);
window.localStorage.setItem(key, value);
```
(of course you should add some sanity checks in there.)

**Comment by Jessekalli** [comment]
+1

**Comment by Awendel** [comment]
+1 for this,

but it should perhaps also be extended for indexedDB.

Since the session class is already associated with such data, it makes a lot of sense to have these available from the Main.

Common use cases would be an OfflineStore so that client has a cached version of all data that can be then synced with the server, and acts as a single source of truth for all associated BrowserWindows / Views.

A pattern which is quite common for true desktop apps.

Of course one can just wrap an API around .executeJavascript, but this introduces unneccessary overhead and relies on a "stable" BrowserWindow / View that currently doesn't perform any work, otherwise latency might be incurred.

**Comment by dheerajkysetti** [comment]
@pabloleban / @Praveer1981 

One of the constraints of accessing localStorage via webContents.executeJavaScript, is the domain level scope of localstorage. 

If there is a need to access localStorage/SessionStorage irrespective of the current domain of the window/webcontents, there is no straightforward way of achieving this.

**Comment by gingters** [comment]
I would also need this.

An app uses `localStorage` to store persisted settings as well as as transient authentication data.
When restarting electron, the last auth data is still there, although expired, so the web page errors out and prompts the user to re-authenticate. For that it redirects to an IdP (other origin). This app uses electron-tabs, so the main window index.html comes from a `file://` origin and creates other webviews in tabs that have the correct origin.

Use case: We want to remove some but not all entries in the localStorage for a certain origin, before we create any browser window and load the applications web page there.

Using the webviews contents with executeJavascript is not a guaranteed way to access the correct origins local storage.

The main issues without an API to access `localStorage` by origin from the main process is:
We need to check when a tab is loaded and is ready. Only at the very first time we would need to access the web view contents, and check if it is at the correct origin. If we have the correct origin, we need to iterate the storage and clear out only the entries that should be deleted. If storage was empty already, or the redirect to the login happens fast, the origin already switched and we can't access the localStorage of the correct origin.

So, as described in the OP, an API on the main processes session object to access the localStorage (and also the sessionStorage) analog to accessing the cookies, with a filter by origin, would be the ideal way to prepare and layout the storage before the browser window is created or the main document is loaded.

**Comment by radislav** [comment]
If adding `session.localStorage` is possible and not blocked for some security reasons, I think I could pay for this patch. 
Contact me :)

**Comment by alinmateutdev** [comment]
+1

**Comment by SlippStream** [comment]
+1

**Comment by QuentinPhrl** [comment]
+10000000

**Comment by alinmateutdev** [comment]
+1

**Comment by ioRekz** [comment]
+1

**Comment by QuentinPhrl** [comment]
> You can get something from the localstorage in your main with
> 
> ```
> const getFromLocalStorage = key => {
> 	myWindow.webContents.executeJavaScript(`localStorage.getItem(${key})`).then(value => value);
> }
> ```

Like many people I think I've used this solution, but I've just encountered a problem with one of my users who has a problem that leaves me perplexed:

In my application, when I try to retrieve a key from local storage, I get the following error.

``const session = await view.webContents.executeJavaScript(
 `localStorage.getItem('session');`
 )``

_Error: Uncaught SecurityError: Failed to read the ‘localStorage’ property from ‘Window’: Access is denied for this document._

From what I've seen, it looks like there's a problem with chrome's third-party cookies, but I don't see the link with electron's chromium engine, so there shouldn't be any configuration sharing.

If you have any solutions, I'd love to hear them.

Thanks in advance

**Comment by QuentinPhrl** [comment]
Hello.

I'd like to relaunch the topic. Do you have any ideas? 

I've tried everything, but some users are still affected by the access rights issue:

_Error: Uncaught SecurityError: Failed to read the ‘localStorage’ property from ‘Window’: Access is denied for this document._

We really need a solution specific to Electron, even though I understand that this poses quite a few technical problems.
